bluez-async/src/modalias.rs

```diff
@@ -0,0 +1,202 @@
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::fmt::{self, Display, Formatter};
+use std::str::FromStr;
+use thiserror::Error;
+
+/// An error parsing a [`Modalias`] from a string.
+#[derive(Clone, Debug, Error, Eq, PartialEq)]
+#[error("Error parsing modalias string {0:?}")]
+pub struct ParseModaliasError(String);
+
+/// A parsed modalias string.
+///
+/// For now only the USB subtype is supported.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Modalias {
+    pub vendor_id: u16,
+    pub product_id: u16,
+    pub device_id: u16,
+}
+
+impl Display for Modalias {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(
+            f,
+            "usb:v{:04X}p{:04X}d{:04X}",
+            self.vendor_id, self.product_id, self.device_id
+        )
+    }
+}
+
+impl FromStr for Modalias {
+    type Err = ParseModaliasError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        RawModalias::from_str(s)?
+            .try_into()
+            .map_err(|_| ParseModaliasError(s.to_owned()))
+    }
+}
+
+impl TryFrom<RawModalias> for Modalias {
+    type Error = ();
+
+    fn try_from(raw: RawModalias) -> Result<Self, Self::Error> {
+        if raw.subtype != "usb" {
+            return Err(());
+        }
+        Ok(Modalias {
+            vendor_id: u16::from_str_radix(raw.values.get("v").ok_or_else(|| ())?, 16)
+                .map_err(|_| ())?,
+            product_id: u16::from_str_radix(raw.values.get("p").ok_or_else(|| ())?, 16)
+                .map_err(|_| ())?,
+            device_id: u16::from_str_radix(raw.values.get("d").ok_or_else(|| ())?, 16)
+                .map_err(|_| ())?,
```

Comment:
given that RawModAlias is private:

I would probably make a
```
pub struct ModaliasBuilder {
    pub vendor_id: Option<u16>,
    pub product_id: Option<u16>,
    pub device_id: Option<u16>,
}
```
with methods to set the fields.

I would then `impl FromStr for Modalias` in terms of ModaliasBuilder (still looping over things, but avoiding the creating a hashmap).

That said, your implementation in terms of hashmap seems fine, and it's unlikely to be performance sensitive.

Reply:
I found it easier to implement the parsing in two stages. I think this will become more useful if we want to add support for other modalias subtypes in future.